k8s.go

```diff
@@ -0,0 +1,643 @@
+func NewK8sClient(s *Server) (*k8sClient, error) {
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the clientset
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &k8sClient{
+		node:              os.Getenv(NODENAME),
+		server:            s,
+		k8scli:            cs,
+		nodeBgpPeerClient: resources.NewNodeBGPPeerClient(cs),
+		nodeBgpCfgClient:  resources.NewNodeBGPConfigClient(cs),
+	}, nil
+}
+
+func (c *k8sClient) updatePrefix() error {
+	var paths []*svbgptable.Path
+	node, err := c.k8scli.Nodes().Get(c.node, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cidr := node.Spec.PodCIDR
+	path, err := c.server.makePath(cidr, false)
+	log.Debugf("Set prefix: %#v", path)
+	paths = append(paths, path)
+	if err = c.server.updatePrefixSet(paths); err != nil {
+		return err
+	}
+	if _, err := c.server.bgpServer.AddPath("", paths); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *k8sClient) initialNeighborConfigSetting() error {
+	bgpconfig, err := c.getBGPConfig()
+	if err != nil {
+		return err
+	}
+	c.lastBgpconfig = bgpconfig
+	neighborConfigs, err := c.getNeighborConfigs(c.lastBgpconfig)
+	if err != nil {
+		return err
+	}
+	for _, n := range neighborConfigs {
+		if err = c.server.bgpServer.AddNeighbor(n); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getNeighborConfigs returns the complete list of BGP neighbor configuration
+// which the node should peer.
+func (c *k8sClient) getNeighborConfigs(bgpconfig map[string]string) ([]*svbgpconfig.Neighbor, error) {
+	var neighbors []*svbgpconfig.Neighbor
+	if mesh, ok := bgpconfig[GlobalNodeMesh]; ok == false {
+		return nil, errors.New("mesh data not found")
+	} else if mesh == `{"enabled":true}` {
+		log.Debug("ebnable mesh")
+		ns, err := c.server.getMeshNeighborConfigs()
+		if err != nil {
+			return nil, err
+		}
+		neighbors = append(neighbors, ns...)
+	}
+	// --- Global peers ---
+	if ns, err := c.server.getGlobalNeighborConfigs(); err != nil {
+		return nil, err
+	} else {
+		neighbors = append(neighbors, ns...)
+	}
+	// --- Node-specific peers ---
+	if ns, err := c.server.getNodeSpecificNeighborConfigs(); err != nil {
+		return nil, err
+	} else {
+		neighbors = append(neighbors, ns...)
+	}
+	log.Debugf("neighbors=%s", neighbors)
+	return neighbors, nil
+}
+
+// checkBGPConfig checks the difference from the last BGP config information.
+// But in a case of the following changing, return error.
+//   /calico/bgp/v1/host/$NODENAME or /calico/global/as_num
+func (c *k8sClient) checkBGPConfig() error {
+	curBgpconfig, err := c.getBGPConfig()
+	if err != nil {
+		return nil
+	}
+	log.Debugf("checkBGPConfig lastBgpconfig: %s", c.lastBgpconfig)
+	if reflect.DeepEqual(c.lastBgpconfig, curBgpconfig) {
```

Comment:
Unless DeepEqual is much faster than CompareMap, this clause is unnecessary.

Reply:
It may be so. But I don't think I will check it.